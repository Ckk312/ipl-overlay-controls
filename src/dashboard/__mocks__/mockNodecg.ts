import { NodeCGBrowser } from 'nodecg/browser';

export const mockSendMessage = jest.fn();
export let replicants: {[key: string]: unknown} = {};

beforeEach(() => {
    replicants = {};
});

window.nodecg = {
    Replicant(name: string) {
        return {
            get value() {
                return replicants[name];
            },
            set value(newValue: unknown) {
                replicants[name] = newValue;
            },
            on: jest.fn()
        };
    },
    sendMessage: mockSendMessage
} as unknown as NodeCGBrowser;
