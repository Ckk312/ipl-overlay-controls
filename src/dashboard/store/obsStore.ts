import { NodeCGBrowser } from 'nodecg/browser';
import { ObsCredentials, ObsData } from 'schemas';
import { createStore, Store, useStore } from 'vuex';
import cloneDeep from 'lodash/cloneDeep';
import { InjectionKey } from 'vue';

const obsData = nodecg.Replicant<ObsData>('obsData');
const obsCredentials = nodecg.Replicant<ObsCredentials>('obsCredentials');

export const obsReps = [obsData, obsCredentials];

export interface ObsStore {
    obsData: ObsData
    obsCredentials: ObsCredentials
}

export const obsStore = createStore<ObsStore>({
    state: {
        obsData: null,
        obsCredentials: null
    },
    mutations: {
        setState(store, { name, val }: { name: string, val: unknown }): void {
            this.state[name] = cloneDeep(val);
        }
    },
    actions: {
        connect(store, { address, password }: { address: string, password?: string }): void {
            nodecg.sendMessage('connectToObs', { address, password });
        }
    }
});

export const obsStoreKey: InjectionKey<Store<ObsStore>> = Symbol();

export function useObsStore(): Store<ObsStore> {
    return useStore(obsStoreKey);
}
