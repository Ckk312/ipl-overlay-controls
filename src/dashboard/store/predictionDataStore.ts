import { NodeCGBrowser } from 'nodecg/browser';
import { createStore, Store, useStore } from 'vuex';
import cloneDeep from 'lodash/cloneDeep';
import { InjectionKey } from 'vue';
import { Prediction, PredictionStore } from 'schemas';
import { PredictionStatus } from 'types/enums/predictionStatus';

export const predictionStore = nodecg.Replicant<PredictionStore>('predictionStore');

export const predictionReps = [ predictionStore ];

export interface PredictionDataStore {
    predictionStore: PredictionStore;
}

export const predictionDataStore = createStore<PredictionDataStore>({
    state: {
        predictionStore: null,
    },
    mutations: {
        setState(store, { name, val }: { name: string, val: unknown }): void {
            this.state[name] = cloneDeep(val);
        },
    },
    actions: {
        async lockPrediction(store) {
            return patchPrediction(
                store.state.predictionStore.currentPrediction,
                PredictionStatus.LOCKED,
                [ PredictionStatus.ACTIVE ]);
        },
        async cancelPrediction(store) {
            return patchPrediction(
                store.state.predictionStore.currentPrediction,
                PredictionStatus.CANCELED,
                [ PredictionStatus.ACTIVE, PredictionStatus.LOCKED ]);
        },
        async resolvePrediction(store, { winningOutcomeIndex }: { winningOutcomeIndex: number }) {
            if (winningOutcomeIndex > 1 || winningOutcomeIndex < 0) {
                throw new Error(`Cannot resolve prediction with outcome index ${winningOutcomeIndex}.`);
            } else if (!store.state.predictionStore.currentPrediction?.id) {
                throw new Error('No prediction to resolve.');
            }

            return nodecg.sendMessage('patchPrediction', {
                id: store.state.predictionStore.currentPrediction.id,
                status: PredictionStatus.RESOLVED,
                winning_outcome_id: store.state.predictionStore.currentPrediction.outcomes[winningOutcomeIndex].id
            });
        }
    }
});

async function patchPrediction(
    currentPrediction: Prediction,
    status: PredictionStatus,
    allowedForStatuses: PredictionStatus[]
): Promise<void> {
    if (!currentPrediction || !allowedForStatuses.includes(currentPrediction?.status as PredictionStatus)
    ) {
        throw new Error(`Cannot set prediction status at this time. Status must match one of [${allowedForStatuses.join(', ')}]`);
    }

    return nodecg.sendMessage('patchPrediction', {
        id: currentPrediction.id,
        status: status,
    });
}

export const predictionDataStoreKey: InjectionKey<Store<PredictionDataStore>> = Symbol();

export function usePredictionDataStore(): Store<PredictionDataStore> {
    return useStore(predictionDataStoreKey);
}
