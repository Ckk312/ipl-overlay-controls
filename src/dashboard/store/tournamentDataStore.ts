import { RoundStore, TournamentData } from 'schemas';
import { createStore, Store, useStore } from 'vuex';
import cloneDeep from 'lodash/cloneDeep';
import { InjectionKey } from 'vue';
import { ToggleTeamImageRequest } from 'types/messages/tournamentData';

const tournamentData = nodecg.Replicant<TournamentData>('tournamentData');
const roundStore = nodecg.Replicant<RoundStore>('roundStore');

export const tournamentDataReps = [ tournamentData, roundStore ];

export interface TournamentDataStore {
    tournamentData: TournamentData;
    roundStore: RoundStore;
}

export const tournamentDataStore = createStore<TournamentDataStore>({
    state: {
        tournamentData: null,
        roundStore: null
    },
    mutations: {
        setState(store, { name, val }: { name: string, val: unknown }): void {
            this.state[name] = cloneDeep(val);
        },
        setTeamImageHidden(store, { teamId, isVisible }: { teamId: string, isVisible: boolean }): void {
            nodecg.sendMessage('toggleTeamImage', { teamId, isVisible } as ToggleTeamImageRequest);
        }
    },
    actions: {
        async getTournamentData(store, { method, id }: { method: TournamentDataStore, id: string }): Promise<void> {
            return await nodecg.sendMessage('getTournamentData', { method, id });
        },
        async getSmashggEvent(store, { eventId }: { eventId: number }): Promise<void> {
            return await nodecg.sendMessage('getSmashggEvent', { eventId });
        }
    }
});

export const tournamentDataStoreKey: InjectionKey<Store<TournamentDataStore>> = Symbol();

export function useTournamentDataStore(): Store<TournamentDataStore> {
    return useStore(tournamentDataStoreKey);
}
