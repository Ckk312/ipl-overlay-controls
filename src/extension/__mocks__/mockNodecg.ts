import express from 'express';
import last from 'lodash/last';

export type ReplicantChangeHandler = (newValue?: unknown, oldValue?: unknown) => void;
export type MockReplicant = { value: unknown };
type OnFunction = (event: string, handler: ReplicantChangeHandler) => void;

export class MockNodecg {
    replicants: {[key: string]: MockReplicant};
    replicantListeners: {[key: string]: ReplicantChangeHandler};
    messageListeners: {[key: string]: (message?: unknown, cb?: () => void) => void};
    requestHandlers: {[type in 'POST' | 'GET']: {[path: string]: express.RequestHandler}};
    mount: jest.Mock;

    constructor() {
        this.replicants = {};
        this.replicantListeners = {};
        this.messageListeners = {};
        this.requestHandlers = { POST: {}, GET: {} };
        this.mount = jest.fn();
    }

    init(): void {
        const self = this;
        require('../helpers/nodecg').set({
            // @ts-ignore: Just a test, improve mocks as needed.
            Replicant(name: string) {
                const replicantValue: { value: unknown, on: OnFunction } = {
                    value: undefined,
                    on: (event: string, handler: ReplicantChangeHandler) => {
                        self.replicantListeners[name] = handler;
                    }
                };
                self.replicants[name] = replicantValue;
                return replicantValue;
            },
            listenFor: (messageName: string, handler: () => void) => {
                self.messageListeners[messageName] = handler as (message: unknown, cb?: () => void) => void;
            },
            Router: () => ({
                post(path: string, ...handlers: express.RequestHandler[]) {
                    self.requestHandlers['POST'][path] = last(handlers);
                }
            }),
            mount: self.mount,
            log: {
                warn: jest.fn()
            }
        });
    }
}
