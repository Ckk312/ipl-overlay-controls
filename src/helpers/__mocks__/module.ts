export type Module = {[key: string]: (...args: unknown[]) => unknown};
