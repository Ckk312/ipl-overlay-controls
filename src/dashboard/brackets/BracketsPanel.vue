<template>
    <ipl-error-display class="m-b-8" />
    <ipl-message
        v-if="!isSupported || !isConfigured"
        type="warning"
    >
        <template v-if="!isSupported">
            Unsupported source ({{ TournamentDataSourceHelper
                .toPrettyString(tournamentDataStore.tournamentData.meta.source as TournamentDataSource) }})
        </template>
        <template v-else>
            Missing configuration for source '{{ TournamentDataSourceHelper
                .toPrettyString(tournamentDataStore.tournamentData.meta.source as TournamentDataSource) }}'
        </template>
    </ipl-message>
    <template v-else>
        <ipl-space 
            class="m-b-8"
            data-test="loaded-bracket-space"
        >
            <div class="title">Loaded bracket</div>
            <template v-if="bracketStore.bracketData == null">
                No brackets currently loaded.
            </template>
            <template v-else>
                <ipl-data-row
                    label="Event Name"
                    :value="bracketStore.bracketData.eventName"
                />
                <ipl-data-row 
                    label="Bracket Name"
                    :value="bracketStore.bracketData.name"
                />
                <ipl-data-row 
                    v-if="matchGroupNames !== bracketStore.bracketData.name"
                    label="Match Groups"
                    :value="matchGroupNames"
                />
                <ipl-data-row
                    v-if="bracketStore.bracketData.roundNumber != null"
                    label="Round Number"
                    :value="`Round ${bracketStore.bracketData.roundNumber}`"
                />
            </template>
        </ipl-space>
        <ipl-space>
            <ipl-button
                label="Load bracket data"
                data-test="load-bracket-data-button"
                async
                @click="getMatchQuery"
            />
            <match-query-param-input
                v-for="param in bracketQuery"
                :key="param.key"
                :param="param"
                :query="queryResult"
                @change="(key, value) => queryResult[key] = value"
                @parameter-add="key => activeParams.add(key)"
                @parameter-remove="key => activeParams.delete(key)"
                @loading="isLoading = $event"
            />
            <ipl-button
                v-if="bracketQuery.length > 0"
                async
                label="Submit"
                data-test="submit-bracket-query-button"
                class="m-t-8"
                :disabled="!activeParamsFilled || isLoading"
                @click="submitBracketQuery"
            />
        </ipl-space>
    </template>
</template>

<script setup lang="ts">
import type NodeCG from '@nodecg/types';
import { IplButton, IplDataRow, IplMessage, IplSpace } from '@iplsplatoon/vue-components';
import { computed, ref, Ref } from 'vue';
import { MatchQueryParameter, MatchImporter, MatchQueryResult } from '@tourneyview/importer';
import MatchQueryParamInput from './MatchQueryParamInput.vue';
import { useBracketStore } from '../store/bracketStore';
import { useTournamentDataStore } from '../store/tournamentDataStore';
import { Configschema } from 'types/schemas';
import { getMatchImporter } from '../../helpers/BracketHelper';
import { TournamentDataSource, TournamentDataSourceHelper } from 'types/enums/tournamentDataSource';
import IplErrorDisplay from '../components/iplErrorDisplay.vue';

const bracketStore = useBracketStore();
const tournamentDataStore = useTournamentDataStore();
const isSupported = computed(() => ['BATTLEFY', 'SMASHGG'].includes(tournamentDataStore.tournamentData.meta.source));
const isConfigured = computed(() => tournamentDataStore.tournamentData.meta.source !== 'SMASHGG' || (nodecg as NodeCG.ClientAPI<Configschema>).bundleConfig?.smashgg?.apiKey != null);

const matchGroupNames = computed(() => {
    if (bracketStore.bracketData == null) {
        return '';
    }

    return bracketStore.bracketData.matchGroups.reduce((result, group, i) => {
        result += group.name;
        if (i !== bracketStore.bracketData.matchGroups.length - 1) {
            result += ', ';
        }

        return result;
    }, '');
});

const activeParams: Ref<Set<string>> = ref(new Set<string>());
const bracketQuery: Ref<MatchQueryParameter[]> = ref([]);
const queryResult: Ref<Record<string, string | number>> = ref({ });
const isLoading = ref<boolean>(false);

function getImporter(): MatchImporter<MatchQueryResult> {
    return getMatchImporter(
        tournamentDataStore.tournamentData.meta.source as TournamentDataSource,
        (nodecg as NodeCG.ClientAPI<Configschema>).bundleConfig);
}

const activeParamsFilled = computed(() => {
    return Array.from(activeParams.value).every(param => queryResult.value[param] != null);
});

async function getMatchQuery() {
    const importer = getImporter();
    bracketQuery.value = await importer.getMatchQueryOptions(tournamentDataStore.tournamentData.meta.id);
}

async function submitBracketQuery() {
    return nodecg.sendMessage('getBracket', queryResult.value);
}
</script>
