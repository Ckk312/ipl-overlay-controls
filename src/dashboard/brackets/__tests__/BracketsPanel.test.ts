import { TestingPinia, createTestingPinia } from '@pinia/testing';
import BracketsPanel from '../BracketsPanel.vue';
import { useBracketStore } from '../../store/bracketStore';
import { useTournamentDataStore } from '../../store/tournamentDataStore';
import { config, mount } from '@vue/test-utils';
import { IplButton, IplMessage, IplSpace } from '@iplsplatoon/vue-components';
import { mockBundleConfig, mockSendMessage } from '../../__mocks__/mockNodecg';
import { BattlefyImporter } from '@tourneyview/importer';
import MatchQueryParamInput from '../MatchQueryParamInput.vue';

jest.mock('@tourneyview/importer');

describe('BracketsPanel', () => {
    config.global.stubs = {
        IplDataRow: true,
        IplButton: true,
        MatchQueryParamInput: true
    };

    let pinia: TestingPinia;

    beforeEach(() => {
        pinia = createTestingPinia();
        config.global.plugins = [pinia];
        mockBundleConfig.smashgg = { apiKey: 'test-startgg-api-key' };
        useTournamentDataStore().tournamentData = {
            // @ts-ignore
            meta: {
                source: 'BATTLEFY',
                id: 'test-tournament-id'
            }
        };
    });

    describe('support/configuration message', () => {
        it('is visible if an unsupported tournament source is in use', () => {
            useTournamentDataStore().tournamentData = {
                // @ts-ignore
                meta: {
                    source: 'UPLOAD'
                }
            };
            const wrapper = mount(BracketsPanel);

            expect(wrapper.findAllComponents(IplSpace).length).toEqual(0);
            const message = wrapper.getComponent(IplMessage);
            expect(message.vm.type).toEqual('warning');
            expect(message.text()).toEqual('Unsupported source (Uploaded file)');
        });
        
        it('is visible if start.gg is unconfigured', () => {
            mockBundleConfig.smashgg = null;
            useTournamentDataStore().tournamentData = {
                // @ts-ignore
                meta: {
                    source: 'SMASHGG'
                }
            };
            const wrapper = mount(BracketsPanel);

            expect(wrapper.findAllComponents(IplSpace).length).toEqual(0);
            const message = wrapper.getComponent(IplMessage);
            expect(message.vm.type).toEqual('warning');
            expect(message.text()).toEqual('Missing configuration for source \'Smash.gg\'');
        });
    });

    describe('loaded bracket message', () => {
        it('shows the expected message when no bracket is loaded', () => {
            useBracketStore().bracketData = undefined;
            const wrapper = mount(BracketsPanel);

            const space = wrapper.getComponent<typeof IplSpace>('[data-test="loaded-bracket-space"]');
            expect(space.html()).toMatchSnapshot();
        });
        
        it('displays bracket data', () => {
            useBracketStore().bracketData = {
                eventName: 'Test Event',
                name: 'Test Bracket',
                matchGroups: [
                    // @ts-ignore
                    { name: 'Test Bracket' }
                ]
            };
            const wrapper = mount(BracketsPanel);

            const space = wrapper.getComponent<typeof IplSpace>('[data-test="loaded-bracket-space"]');
            expect(space.html()).toMatchSnapshot();
        });
        
        it('displays bracket data when more than one match group is loaded', () => {
            useBracketStore().bracketData = {
                eventName: 'Test Event',
                name: 'Test Bracket',
                matchGroups: [
                    // @ts-ignore
                    { name: 'Group One' },
                    // @ts-ignore
                    { name: 'Group Two' }
                ]
            };
            const wrapper = mount(BracketsPanel);

            const space = wrapper.getComponent<typeof IplSpace>('[data-test="loaded-bracket-space"]');
            expect(space.html()).toMatchSnapshot();
        });
        
        it('displays bracket data when a round number is present', () => {
            useBracketStore().bracketData = {
                eventName: 'Test Event',
                name: 'Test Bracket',
                matchGroups: [
                    // @ts-ignore
                    { name: 'Test Bracket' }
                ],
                roundNumber: 8
            };
            const wrapper = mount(BracketsPanel);

            const space = wrapper.getComponent<typeof IplSpace>('[data-test="loaded-bracket-space"]');
            expect(space.html()).toMatchSnapshot();
        });
    });

    describe('loading bracket data', () => {
        it('handles loading and editing bracket query parameters and submitting a query', async () => {
            (BattlefyImporter.prototype.getMatchQueryOptions as jest.Mock).mockResolvedValue([
                { type: 'static', value: 'test', key: 'testParam' }
            ]);
            const wrapper = mount(BracketsPanel);

            expect(wrapper.findAllComponents('match-query-param-input-stub').length).toEqual(0);

            await wrapper.getComponent<typeof IplButton>('[data-test="load-bracket-data-button"]').trigger('click');

            expect(BattlefyImporter.prototype.getMatchQueryOptions).toHaveBeenCalledWith('test-tournament-id');
            const paramInputs = wrapper.findAllComponents<typeof MatchQueryParamInput>('match-query-param-input-stub');
            expect(paramInputs.length).toEqual(1);
            const paramInput = paramInputs.at(0);
            paramInput.vm.$emit('parameter-add', 'testParam');
            paramInput.vm.$emit('change', 'testParam', 'testParamValue');

            await wrapper.getComponent<typeof IplButton>('[data-test="submit-bracket-query-button"]').trigger('click');
            expect(mockSendMessage).toHaveBeenCalledWith('getBracket', { testParam: 'testParamValue' });
        });

        it('disables submitting a query if data is loading', async () => {
            (BattlefyImporter.prototype.getMatchQueryOptions as jest.Mock).mockResolvedValue([
                { type: 'static', value: 'test', key: 'testParam' }
            ]);
            const wrapper = mount(BracketsPanel);

            await wrapper.getComponent<typeof IplButton>('[data-test="load-bracket-data-button"]').trigger('click');
            const paramInputs = wrapper.findAllComponents<typeof MatchQueryParamInput>('match-query-param-input-stub');
            expect(paramInputs.length).toEqual(1);
            const paramInput = paramInputs.at(0);

            paramInput.vm.$emit('loading', true);
            await wrapper.vm.$nextTick();
            expect(wrapper.getComponent<typeof IplButton>('[data-test="submit-bracket-query-button"]').vm.disabled).toEqual(true);
            
            paramInput.vm.$emit('loading', false);
            await wrapper.vm.$nextTick();
            expect(wrapper.getComponent<typeof IplButton>('[data-test="submit-bracket-query-button"]').vm.disabled).toEqual(false);
        });

        it('disables submitting a query if data is missing', async () => {
            (BattlefyImporter.prototype.getMatchQueryOptions as jest.Mock).mockResolvedValue([
                { type: 'static', value: 'test', key: 'testParam' }
            ]);
            const wrapper = mount(BracketsPanel);

            await wrapper.getComponent<typeof IplButton>('[data-test="load-bracket-data-button"]').trigger('click');
            const paramInputs = wrapper.findAllComponents<typeof MatchQueryParamInput>('match-query-param-input-stub');
            expect(paramInputs.length).toEqual(1);
            const paramInput = paramInputs.at(0);

            paramInput.vm.$emit('parameter-add', 'testParam');
            await wrapper.vm.$nextTick();
            expect(wrapper.getComponent<typeof IplButton>('[data-test="submit-bracket-query-button"]').vm.disabled).toEqual(true);

            paramInput.vm.$emit('change', 'testParam', 'testParamValue');
            await wrapper.vm.$nextTick();
            expect(wrapper.getComponent<typeof IplButton>('[data-test="submit-bracket-query-button"]').vm.disabled).toEqual(false);
        });
    });
});
