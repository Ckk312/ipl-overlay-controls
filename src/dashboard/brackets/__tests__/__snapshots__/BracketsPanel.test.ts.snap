// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BracketsPanel loaded bracket message displays bracket data 1`] = `
"<div data-v-d255ffee="" class="ipl-space color-dark m-b-8" data-test="loaded-bracket-space">
  <div class="title">Loaded bracket</div>
  <ipl-data-row-stub label="Event Name" copiable="false" value="Test Event"></ipl-data-row-stub>
  <ipl-data-row-stub label="Bracket Name" copiable="false" value="Test Bracket"></ipl-data-row-stub>
  <!--v-if-->
  <!--v-if-->
</div>"
`;

exports[`BracketsPanel loaded bracket message displays bracket data when a round number is present 1`] = `
"<div data-v-d255ffee="" class="ipl-space color-dark m-b-8" data-test="loaded-bracket-space">
  <div class="title">Loaded bracket</div>
  <ipl-data-row-stub label="Event Name" copiable="false" value="Test Event"></ipl-data-row-stub>
  <ipl-data-row-stub label="Bracket Name" copiable="false" value="Test Bracket"></ipl-data-row-stub>
  <!--v-if-->
  <ipl-data-row-stub label="Round Number" copiable="false" value="Round 8"></ipl-data-row-stub>
</div>"
`;

exports[`BracketsPanel loaded bracket message displays bracket data when more than one match group is loaded 1`] = `
"<div data-v-d255ffee="" class="ipl-space color-dark m-b-8" data-test="loaded-bracket-space">
  <div class="title">Loaded bracket</div>
  <ipl-data-row-stub label="Event Name" copiable="false" value="Test Event"></ipl-data-row-stub>
  <ipl-data-row-stub label="Bracket Name" copiable="false" value="Test Bracket"></ipl-data-row-stub>
  <ipl-data-row-stub label="Match Groups" copiable="false" value="Group One, Group Two"></ipl-data-row-stub>
  <!--v-if-->
</div>"
`;

exports[`BracketsPanel loaded bracket message shows the expected message when no bracket is loaded 1`] = `
"<div data-v-d255ffee="" class="ipl-space color-dark m-b-8" data-test="loaded-bracket-space">
  <div class="title">Loaded bracket</div> No brackets currently loaded.
</div>"
`;
